src/coreclr/jit/lsraarm64.cpp

```diff
@@ -2040,26 +2040,16 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree, int* pDstCou
                 buildInternalIntRegisterDefForNode(embOp2Node);
             }
 
-            size_t prefUseOpNum = 1;
-            if (intrinEmb.id == NI_Sve_CreateBreakPropagateMask)
-            {
-                prefUseOpNum = 2;
-            }
-            GenTree* prefUseNode = embOp2Node->Op(prefUseOpNum);
+            // The embedded op may be prefixed by a MOVPFRX instruction. If so, then the embedded
+            // op can't use the reuse a source register as the destination. Ensure all the inputs
+            // are marked as delay free.
```

Comment:
We shouldn't be delay freeing everything right, because aren't we are expecting that `op1` and the `destReg` be the same to avoid additional unnecessary moves?

That is, take `add (Vectors, predicated)` for example. In that case we have:
```
ADD <Zdn>.<T>, <Pg>/M, <Zdn>.<T>, <Zm>.<T>
```

So in the scenario that we don't need `movprfx`, then `tgtPrefUse == op1` (therefore `op1` is not delay free) and `op2` does not need to be delay free. However, in the case we do need `movprfx`, then we still want `tgtPrefUse == op1` (and therefore `op1` is not delay free) but we do require that `op2` be delay free, due to the `movprfx` requirements. -- If we're marking `op1` as delay free, then it won't be assign the same register as the destination, and that means we need to insert an additional `mov tgt, op1` which should be preventing us from emitting `movprfx` anyways (and we should be emitting `add (unpredicated); sel` instead).

For .NET 9, we likely want to pessimize and assume that `movprfx` will also be needed, just because that's the simpler fix and we're already in RC2. However, we should ensure that an issue exists for .NET 10 and we fix the codegen to do the optimal thing here before we ship the feature as stable.

That includes handling (in .NET 10) cases like `add` which are commutative, where we want to determine which of `op1` or `op2` is more likely to be selected as the destination (normally done via `getTgtPrefOperands` or similar) and swap them if `op2` is more likely. -- That is, I imagine we really want something like xarch's `LinearScan::BuildRMWUses` for the SVE RMW instructions here.

Reply:
> For .NET 9, we likely want to pessimize and assume that `movprfx` will also be needed, just because that's the simpler fix and we're already in RC2. However, we should ensure that an issue exists for .NET 10 and we fix the codegen to do the optimal thing here before we ship the feature as stable.

Agreed and will do.


> We shouldn't be delay freeing everything right, because aren't we are expecting that `op1` and the `destReg` be the same to avoid additional unnecessary moves?

> However, in the case we do need `movprfx`, then we still want `tgtPrefUse == op1` (and therefore `op1` is not delay free) but we do require that `op2` be delay free, due to the `movprfx` requirements.

Which means the original code was doing the right thing. `Op1` used `BuildUse()` and `Op2` used `BuildDelayFreeUses()`.

But, the jit was still allowing `Op1` and `Op2` to have the same registers.

With the code in HEAD we get:

```
Generating: N029 (  4,  3) [000026] ---XG+-----                   t26 = *  IND       simd16 REG d16 <l:$2c1, c:$2c2>
IN000a:             ldr     q16, [x0, #0x08]
                                                                        /--*  t26    simd16 
Generating: N031 (  4,  3) [000052] DA-XG------                         *  STORE_LCL_VAR simd16 V07 cse0         d:1 d16 REG d16 $VN.Void
							V07 in reg d16 is becoming live  [000052]
							Live regs: 0000000000000001 {x0} + {d16} => 0001000000000001 {x0 d16}
							Live vars after [000052]: {V03} +{V07} => {V03 V07}
Generating: N033 (  1,  1) [000053] -----------                   t53 =    LCL_VAR   simd16 V07 cse0         u:1 d16 REG d16 <l:$2c1, c:$2c2>
Generating: N035 (  1,  1) [000055] -----------                   t55 =    LCL_VAR   simd16 V07 cse0         u:1 d16 (last use) REG d16 <l:$2c0, c:$300>
Generating: N037 (???,???) [000073] -----------                   t73 =    HWINTRINSIC mask   short CreateTrueMaskAll REG p0
IN000b:             ptrue   p0.h
Generating: N039 (???,???) [000074] -c---------                   t74 =    CNS_VEC   simd16<0x00000000, 0x00000000, 0x00000000, 0x00000000> REG NA
                                                                        /--*  t53    simd16 
                                                                        +--*  t55    simd16 
Generating: N041 (  7,  6) [000034] -A-XG+-----                   t34 = *  HWINTRINSIC simd16 short Max REG NA <l:$2c4, c:$2c5>
                                                                        /--*  t73    mask   
                                                                        +--*  t34    simd16 
                                                                        +--*  t74    simd16 
Generating: N043 (???,???) [000075] -A-XG------                   t75 = *  HWINTRINSIC simd16 short ConditionalSelect REG d16
							V07 in reg d16 is becoming dead  [000055]
							Live regs: 0001000000000001 {x0 d16} - {d16} => 0000000000000001 {x0}
							Live vars after [000055]: {V03 V07} -{V07} => {V03}
```

So, destination of `ConditionalSelect+Max` is `d16`, and both inputs to `Max` are `d16`.


This can be fixed by removing the `prefUseNode` from the call to `BuildDelayFreeUses()` - see latest version of the PR. Now what happens is:

```
Generating: N029 (  4,  3) [000026] ---XG+-----                   t26 = *  IND       simd16 REG d16 <l:$2c1, c:$2c2>
IN000a:             ldr     q16, [x0, #0x08]
                                                                        /--*  t26    simd16 
Generating: N031 (  4,  3) [000052] DA-XG------                         *  STORE_LCL_VAR simd16 V07 cse0         d:1 d16 REG d16 $VN.Void
							V07 in reg d16 is becoming live  [000052]
							Live regs: 0000000000000001 {x0} + {d16} => 0001000000000001 {x0 d16}
							Live vars after [000052]: {V03} +{V07} => {V03 V07}
Generating: N033 (  1,  1) [000053] -----------                   t53 =    LCL_VAR   simd16 V07 cse0         u:1 d16 REG d16 <l:$2c1, c:$2c2>
Generating: N035 (  1,  1) [000055] -----------                   t55 =    LCL_VAR   simd16 V07 cse0         u:1 d16 (last use) REG d16 <l:$2c0, c:$300>
Generating: N037 (???,???) [000073] -----------                   t73 =    HWINTRINSIC mask   short CreateTrueMaskAll REG p0
IN000b:             ptrue   p0.h
Generating: N039 (???,???) [000074] -c---------                   t74 =    CNS_VEC   simd16<0x00000000, 0x00000000, 0x00000000, 0x00000000> REG NA
                                                                        /--*  t53    simd16 
                                                                        +--*  t55    simd16 
Generating: N041 (  7,  6) [000034] -A-XG+-----                   t34 = *  HWINTRINSIC simd16 short Max REG NA <l:$2c4, c:$2c5>
                                                                        /--*  t73    mask   
                                                                        +--*  t34    simd16 
                                                                        +--*  t74    simd16 
Generating: N043 (???,???) [000075] -A-XG------                   t75 = *  HWINTRINSIC simd16 short ConditionalSelect REG d17
							V07 in reg d16 is becoming dead  [000055]
							Live regs: 0001000000000001 {x0 d16} - {d16} => 0000000000000001 {x0}
							Live vars after [000055]: {V03 V07} -{V07} => {V03}
IN000c:             movprfx z17.h, p0/z, z16.h
IN000d:             smax    z17.h, p0/m, z17.h, z16.h
```

The same, but destination of `ConditionalSelect+Max` is now `d17`. The jit can now do the `movprfx` from `d16` to `d17` and also reuse `d16` in the `smax`.

The final assembly is correct. However what I'm unsure about is that if the lsra code is now correct, or if it's a fluke that it works.

(For reference, [failing jit dump](https://gist.github.com/a74nh/d094c16c57166195302ff7cdb53f2d1f) and [fixed jit dump](https://gist.github.com/a74nh/552e52f1e5692530a138232b276592c3))

Reply:
The ref positions created for both uses of `V07` are marked as last uses which is a bit surprising to me. It means that we will free the register at this location and also at the next location.
Even if they weren't, if we end up with the non-delay free ref position as the last use, it would end up freeing the register at this location. I don't know whether that is possible or not; I don't really follow the logic that builds these ref positions.
It's possible the main allocation logic here has a bug and that it needs to be paying special attention to `regsInUseNextLocation` as part of adding to the `regsToFree` set. However, both RPs getting `last` is definitely strange.

The safest fix for now would definitely be to consider all operands as delay freed.

Reply:
One thing I don't quite follow: why can we preference the target to be `op1` if we are going to emit `movprfx` that does not allow the destination register to be the same as `op1`?

Reply:
> The ref positions created for both uses of `V07` are marked as last uses which is a bit surprising to me.

Ah, that's a red herring, it's simply because when we print the first RP it is the last one, so at that point it is marked "last".

So indeed, passing `rmwNode` to `BuildDelayFreeUses` in this case makes it skip setting that ref position as delay-freed. I think that is an optimization; for example, it allows `x = x + x` to end up as `add rax, rax` on xarch. Without the optimization the destination would not be allowed to be `rax`.
But I suppose this optimization is just illegal in this case.

The only thing I'm unsure about is whether or not the operand we pick as the "delay" freed one is always going to be the last operand. Otherwise we would run into the problem I described above, where the ref position marked "last" is not the delay-freed one, causing us to free the register at the use's location.

Reply:
> One thing I don't quite follow: why can we preference the target to be op1 if we are going to emit movprfx that does not allow the destination register to be the same as op1?

The encoding requires `dst == op1`, because its RMW. You can't individually encode the separate registers. The `movprfx` then further requires that `op2 != dst`, there is no allowance for `dst == op1 == op2`.

Reply:
The simpler fix may be to handle `dst == op1 == op2` and emit the `ins; sel` pattern instead of `movprfx; sel`. Or to just never emit `mvprfx` in .NET 9

But if there is an easy way to guarantee that `dst != op2` when we know `movprfx` will be emitted (or we pessimize to assume it will for the .NET 9 fix), then that is the more ideal long term solution.

Reply:
> But if there is an easy way to guarantee that `dst != op2`

Is that not covered by simply not passing `prefUseNode` to `BuildDelayFreeUses()` - as it's done in the latest version of the PR?
`srcCount += BuildDelayFreeUses(embOp2Node->Op(argNum));`

I do think there needs some better testing around this area, which is why I've added some extra hwintrinsic tests as part of #107084

Reply:
Yes, the delay free should accomplish it. The only concern I have is whether the delay free will always end up taking effect. I think you should audit the code and verify by inspection that, if the operands refer to the same interval (local), then the last ref position created will always be the one to be marked delay-free. A comment that there is a subtle requirement about that would be good too.

Also, it would be a good idea to inspect the other places that build delay free uses because of hard requirements to verify that they do not have this same bug around `rmwNode`.

Reply:
Because we're not passing in a rmwnode to `BuildDelayFreeUses()` then the op2, op3 etc will always be marked as delay free.
I've added some assert checks to confirm this.
I've also merged it with the very similar code for FMA to remove duplication.

I'm not sure what would happen if the two locals were passed to the hwintrinsic in a different order, eg `55` then `52`
```
[000053] -----------                   t53 =    LCL_VAR   simd16 V07 cse0         u:1 d16 REG d16 <l:$2c1, c:$2c2>
[000055] -----------                   t55 =    LCL_VAR   simd16 V07 cse0         u:1 d16 (last use) REG d16 <l:$2c0, c:$300>
                                             /--*  t55    simd16 
                                             +--*  t53    simd16 
[000034] -A-XG+-----                   t34 = *  HWINTRINSIC simd16 short Max REG NA <l:$2c4, c:$2c5>
```
But I'm not sure that's possible.

Reply:
The order of them does not matter -- LSRA computes the precise last use information based on the ref positions, hence the last ref position created will be the one to be considered the "last use". For a last use LSRA will then free the register at location N + 1, or if the ref position is also marked as delay-freed, at location N + 2. Every node generally has its uses happen at location N and its definition happen at location N + 1.
So the only problem would be if we ended up creating the ref positions such that we ended up with two ref positions 
```
RP <not last> <delay free>
RP <last> <not delay free>
```

I think LSRA would free the register at location N + 1 in this case. To be honest that seems like a bug in the allocation algorithm, but not one I would be comfortable fixing for .NET 9 at this stage.

I think we are good with the fix to just not pass `rmwNode` now that you've went through some of the other cases and merged the code.

Reply:
Opened #107134 to improve this code

Reply:
I will sync-up with @a74nh offline about the general problem and right way to fix it. For now, the current fix sounds reasonable.